### ISSUE TITLE
[FEATURE] Add RFC 8760 Support (SHA-256 and SHA-512/256 digest auth)

### ISSUE THREAD

**Issue by lanxiaolin-sz** [OpenSIPS/opensips#2227]
**When the client uses the user password for digest authentication, MD5 will be used for encryption; at present, MD5 has proven to be insecure enough. Can we consider using SHA256 instead of MD5?**

**Replace MD5 with SHA256 [auth module]**

**Implementation**
 
  - Component: auth module、auth_db module 
  - Type: function
  - Name: challenge

**Replace MD5 with SHA256**

**If there are already available implementations, please let me know.**

**Comment by liviuchircu** [comment]
Hey, @lanxiaolin-sz!

Thank you for your interest!  @sobomax already started a discussion on this topic [here](https://groups.google.com/g/sip-chronicles/c/cbHjLkojfmc), which materialized into SIP Chronicles [episode 9](https://www.youtube.com/watch?v=SoEHVvypUB0).

He is already working on a prototype implementation of RFC 8760, which he aims to present during the OpenSIPS Summit Wokshops, Monday 14th September.

**Comment by sobomax** [comment]
Thanks Liviu, working on it. Should we perhaps keep this issue open to track and co-ordinate progress? The change eventually would touch many parts I think ("HA1b" effect). Given the amount of time I am not bold enough to promise a full-blown solution I think. :)

-Max

**Comment by sobomax** [comment]
Hey Liviu, making some progress with a python prototype, it seems now quite RFC8760 complete, check this out:

`00:00:06.313/GLOBAL/bob_ua: SENDING message to 127.0.0.1:5060:
SIP/2.0 401 Unauthorized
Via: SIP/2.0/UDP 127.0.0.1:5060;branch=z9hG4bKdb8464cc9830d88962a3d5c38d4dad3b;rport=5060
From: "Alice Smith" <sip:alice_reinv_brkn2_ipv4@127.0.0.1>;tag=36b0297d58addb020ab99ec788779e88
To: <sip:bob_reinv_brkn2@127.0.0.1>;tag=9d86c7b873d19ca3d1834ab463cb3185
Call-ID: [}SV"*\Q!UG%Q8\d%vV6:<2_IUJHQ>`A@wZ"2kvj7i{Z3:L/m-b2b_1
CSeq: 200 INVITE
WWW-Authenticate: Digest realm="127.0.0.1",nonce="wObU+nYqXknWs2LrnB5jrHHlRnM+7ZAKlOPyAd5HGDQ",qop=auth,algorithm=SHA-512-256
WWW-Authenticate: Digest realm="127.0.0.1",nonce="wObU+nYqXknWs2LrnB5jrHHlRnM+7ZAKlOPyAd5HGDQ",qop=auth,algorithm=SHA-256
WWW-Authenticate: Digest realm="127.0.0.1",nonce="wObU+nYqXknWs2LrnB5jrHHlRnM+7ZAKlOPyAd5HGDQ",qop=auth,algorithm=MD5
Content-Length: 0

00:00:06.381/GLOBAL/alice_ua: SENDING message to 127.0.0.1:5060:
INVITE sip:bob_reinv_brkn2@127.0.0.1:5060 SIP/2.0
Via: SIP/2.0/UDP 127.0.0.1:5061;branch=z9hG4bK6b4ff604325a4986618a79ebab1a49e1;rport
Max-Forwards: 70
From: "Alice Smith" <sip:alice_reinv_brkn2_ipv4@127.0.0.1>;tag=iFGdrB.SiI.z+xS4~wg0DG+uC.2JyOdl
To: <sip:bob_reinv_brkn2@127.0.0.1>
Call-ID: [}SV"*\Q!UG%Q8\d%vV6:<2_IUJHQ>`A@wZ"2kvj7i{Z3:L/m
CSeq: 201 INVITE
Contact: Anonymous <sip:alice_reinv_brkn2_ipv4@127.0.0.1:5061>
Expires: 300
User-Agent: Sippy
Authorization: Digest username="mightyuser",realm="127.0.0.1",nonce="wObU+nYqXknWs2LrnB5jrHHlRnM+7ZAKlOPyAd5HGDQ",uri="sip:bob_reinv_brkn2@127.0.0.1:5060",response="938405e364c7c22045fe8806c03d47ce53a08e3214e60fa72df7ff2d689118cd",algorithm=SHA-512-256,qop=auth,nc=00000001,cnonce="2bbf01ad"
Content-Type: application/sdp
Content-Length: 405`

I'd appreciate if you can take a quick peek with your sharp eye and see if you can spot any things that are against that RFC.  Full logs are here if you are curious:

https://travis-ci.com/github/sippy/voiptests/builds/182659671

It would also be cool if you can independently hand-verify my usage of the SHA256 by calculating the response (using scripting language of your choice), the password in this case  's3cr3tpAssw0Rd' (w/o quotes).

**Comment by liviuchircu** [comment]
Hey, Max!

No obvious issues with the syntax.  Well done, looking good!

However, it seems my `response=` differs from yours, so I will list my computation steps for both SHA-256 and SHA-512/256:

## SHA-512/256

```
1. HA1 = H(username:realm:password)
       = H(mightyuser:127.0.0.1:s3cr3tpAssw0Rd)
       = 8e845b39382c4cd94f9ac6cf873b423b0b69a6b444f9fea8a3e7f75ab4694d67

2.  HA2 = H(method:digestURI)
       = H(INVITE:sip:bob_reinv_brkn2@127.0.0.1:5060)
       = bffae99cce418feb850aa1d7aba79ec6341307f609b8fa71391628ad01694b48

3.  response = H(HA1:nonce:nonceCount:cnonce:qop:HA2)
       = H(8e845b39382c4cd94f9ac6cf873b423b0b69a6b444f9fea8a3e7f75ab4694d67:wObU+nYqXknWs2LrnB5jrHHlRnM+7ZAKlOPyAd5HGDQ:00000001:2bbf01ad:auth:bffae99cce418feb850aa1d7aba79ec6341307f609b8fa71391628ad01694b48)
       = 746e583fabd49f36018c2dc426fc583b7d618704e928c05825bc2ddf0721b8ef

## SHA-256

```
1. HA1 = H(username:realm:password)
       = H(mightyuser:127.0.0.1:s3cr3tpAssw0Rd)
       = 282e911e51719d3ca528447b3b75d451668edfa4cdc00f34844a9f3f00d0b3d5

2.  HA2 = H(method:digestURI)
       = H(INVITE:sip:bob_reinv_brkn2@127.0.0.1:5060)
       = e9657afb54da6e325eaa07dd7673968483be34023d3dbcc0c1e29b5ac7a30aac

3.  response = H(HA1:nonce:nonceCount:cnonce:qop:HA2)
       = H(282e911e51719d3ca528447b3b75d451668edfa4cdc00f34844a9f3f00d0b3d5:wObU+nYqXknWs2LrnB5jrHHlRnM+7ZAKlOPyAd5HGDQ:00000001:2bbf01ad:auth:e9657afb54da6e325eaa07dd7673968483be34023d3dbcc0c1e29b5ac7a30aac)
       = 01833ce1d7cc53a459ff146317e969d50fab6fd9101d19fab10b9023c3286f2f
```

Let me know your thoughts.  Don't exclude the possibility of me doing something terribly wrong here, I'd like to learn about your algorithm's steps, so we can compare!

After all, I did learn this from [the wiki](https://en.wikipedia.org/wiki/Digest_access_authentication), lol!

**Comment by sobomax** [comment]
Hey Liviu, do you want to post your changes made @opensipit into some branch perhaps? I'd really like get it merged with  my branch here https://github.com/OpenSIPS/opensips/compare/master...sippy:RFC8760. It would serve as a good indication where SHA-256/SHA-512-256 needs to be added, so I don't have to dig it up. :)

**Comment by liviuchircu** [comment]
hey @sobomax!  I've already pushed the fix to the `3.1` stable branch, see 27d5862fb.  However, I want it to receive at least 1 month of testing before backporting it to `3.0` or `2.4`, since lots of deployments are currently using those versions and it's not worth risking a regression.

**Comment by liviuchircu** [comment]
Time flies so fast, we didn't even realize to close this issue once it was completed!  I will happily do so ASAP, cheers folks!

@lanxiaolin-sz: although the feature is ready, on the provisioning side we still need to update the `opensips-cli` to fill in the new `subscriber.ha1_sha256` and `subscriber.ha1_sha512t256` columns.  If you do not have the tooling to generate those hashes on your own, stay tuned for updates on the [opensips-cli](https://github.com/opensips/opensips-cli) project.
